compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp

```diff
@@ -103,28 +105,53 @@ makeEncoding(OpBuilder &builder, IREE::LinalgExt::EncodingUser user,
       context, userAttr, roleAttr, operandElemTypesAttr, originalTypeAttr);
 }
 
+static Value castEncodedResult(OpBuilder &builder, Location loc, Value encoded,
+                               Value castResult,
+                               IREE::LinalgExt::EncodingAttr encodingAttr) {
+  auto encodedType = cast<RankedTensorType>(encoded.getType());
+  auto castResultType = cast<RankedTensorType>(castResult.getType());
+  auto castedType = RankedTensorType::get(
+      encodedType.getShape(), castResultType.getElementType(), encodingAttr);
+  Operation *castOp = getDefiningCastOp(castResult);
+  return builder
+      .create(loc, castOp->getName().getIdentifier(), encoded, castedType,
+              castOp->getAttrs())
+      ->getResult(0);
+}
+
 static Value padAndSetEncoding(OpBuilder &builder, Location loc, Value source,
                                IREE::LinalgExt::EncodingUser user,
                                IREE::LinalgExt::EncodingRole role,
-                               TypeRange operandTypes) {
+                               TypeRange operandTypes,
+                               bool castResult = false) {
+  Value padSource = source;
+  if (castResult) {
+    auto castOp = source.getDefiningOp();
+    padSource = castOp->getOperand(0);
+  }
```

Comment:
If we replace `bool castResult` with `std::optional<CastOpInterface> srcCastOp`, we can simplify it to `Value padSource = srcCastOp ? srcCastOp->getOperand(0) ? source;`